### ISSUE TITLE
IA Uploader causes query that fills our hard drive with temp files (woah)

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#901]
Over the weekend things hit a bit of a crisis point due to some changes I made last week to the IA uploader to fix #899. The changes I made tweaked the query that was used. That caused it to use a hash join. 

I don't know anything about hash joins, but I learned on Sunday that they're very bad. This one, in particular, required the generation of about 200GB of temporary files. Those files filled the hard drive and caused the crisis. I was able to mitigate the crisis by staying ahead of things as the disk filled, because at first I couldn't get enough analysis done to figure out the cause, so I just had to delete large things to make space. This was not fun. 

I knew Django's `Prefetch` mechanism made some nasty queries, but I never expected a query to require 200GB of temporary tables to run. Follows are some notes on what I know about this query. For now the IA uploader is down — that stopped the problem from coming back.

Below are some additional notes until I can get back to this issue. Many of these come from the superbly helpful folks on the postgresql IRC channel:

1. `log_temp_files` is a way of logging temporary file metadata for later analysis.

    > https://www.postgresql.org/docs/9.5/static/runtime-config-logging.html

1. The query itself:

    - Was logged to the slow query log and took about 43 minutes (2579861.432ms to be precise).
    - Can be found `EXPLAIN`ed here (in abridged form): https://explain.depesz.com/s/MZhl

1. The slow query log from the time of the incident is here: [slow_queries.log](https://github.com/freelawproject/courtlistener/files/2555622/slow_queries.log)

1. More research about slow hash joins is needed, but the real solution is probably disabling the prefetch somehow. I suspect it is faster for small dockets and slower for big ones. Is it worth investigating the threshold and figuring out a way to do the right thing?

**Comment by mlissner** [comment]
Unfortunately, this issue may have to wait until next week or maybe later for some love from me. If anybody else has time or energy to look into this, I'd welcome their taking a look.

**Comment by mlissner** [comment]
A dumb fix for this might be to just point the uploader to our read replica until the queries can be analyzed properly. I'm tempted to do that now, but I want to get a handle on our AWS costs before starting to run all these queries. Still, it's a nice, simple solution.

**Comment by mlissner** [comment]
There's a note in the "Crypto" section that says:

> If you have recently signed up for Cloudflare, and your certificate status above shows “Authorizing Certificate”, HTTPS is not yet available for your site because Cloudflare does not have a certificate for it. Provisioning typically takes around 15 minutes for paid plans and up to 24 hours for Free. 

I don't know what starts that clock, but it sort of sounds like we'll have downtime when we switch of up to 24 hours. That's be total shite.

**Comment by mlissner** [comment]
OK, I've been working on this issue for a few days now. It's a tough one for me because it requires mastery of both Django's ORM and postgresql's query planner. I've been fortunate to receive a bucketload of help from both the postgresql and django IRC channels, and I think I've gotten to the bottom of this. 

Here's the query that does bad things, slightly simplified for posting here:

```sql
SELECT DISTINCT 
  ("people_db_role"."party_id") AS "_prefetch_related_val_party_id", 
  "people_db_attorney"."id"
FROM 
  "people_db_attorney" 
INNER JOIN "people_db_role" ON 
  ("people_db_attorney"."id" = "people_db_role"."attorney_id" ) 
INNER JOIN "people_db_role" T4 ON 
  ("people_db_attorney"."id" = T4."attorney_id" ) 
WHERE 
  ("people_db_role"."docket_id" = 7926834 AND T4."party_id" IN (2374386, 2374387))
```

The simplification is that that final `IN` statement only has two params instead of a few thousand. Now, if you look at the joins, you'll see something weird: It's joining on the same table with the same join condition twice. That's not good. 

When you do that, and you run an `EXPLAIN ANALYZE` statement for the above query, you get (roughly, this is copy pasting across a few day's worth of these and might not be the exact one):

```txt
Unique  (cost=1591.96..1591.98 rows=1 width=199) (actual time=541.230..673.842 rows=7361 loops=1)
   ->  Sort  (cost=1591.96..1591.96 rows=1 width=199) (actual time=541.227..568.385 rows=163224 loops=1)
         Sort Key: people_db_role.party_id, people_db_attorney.id, people_db_attorney.date_created, people_db_attorney.date_modified, people_db_attorney.name, people_db_attorney.contact_raw, people_db_attorney.phone, people_db_attorney.fax, people_db_attorney.email
         Sort Method: quicksort  Memory: 50681kB
         ->  Nested Loop  (cost=1.43..1591.95 rows=1 width=199) (actual time=0.621..139.737 rows=163224 loops=1)
               ->  Nested Loop  (cost=0.99..1181.94 rows=137 width=199) (actual time=0.361..1.345 rows=72 loops=1)
                     ->  Index Only Scan using people_db_role_party_id_73709a165082400a_uniq on people_db_role t4  (cost=0.56..24.63 rows=137 width=4) (actual time=0.343..0.694 rows=72 loops=1)
                           Index Cond: (party_id = ANY ('{2374386,2374387,2374388,2374389}'::integer[]))
                           Heap Fetches: 0
                     ->  Index Scan using people_db_attorney_pkey on people_db_attorney  (cost=0.43..8.45 rows=1 width=195) (actual time=0.007..0.007 rows=1 loops=72)
                           Index Cond: (id = t4.attorney_id)
               ->  Index Scan using people_db_role_9f99f769 on people_db_role  (cost=0.43..2.98 rows=1 width=8) (actual time=0.015..1.505 rows=2267 loops=72)
                     Index Cond: (attorney_id = people_db_attorney.id)
                     Filter: (docket_id = 7926834)
 Planning time: 1.097 ms
 Execution time: 679.646 ms
```

If you prefer a visual version of that you can find [one here][bad] (also pasted below for posterity):

![screenshot from 2019-01-03 22-45-43](https://user-images.githubusercontent.com/236970/50676464-53b87d00-0fa9-11e9-9a44-cbdb862e2df0.png)

Note how at one point it does a sort over 163k rows. That's in order to do a `DISTINCT` on them, and as you can imagine, that's very bad. Imagine how much worse it gets when there's more than a few people IDs provided in the query.

-------

That's the status quo we're out to fix. Now, here's a new query, except removing the double join business:

```sql
SELECT DISTINCT 
  ("people_db_role"."party_id") AS "_prefetch_related_val_party_id", 
  "people_db_attorney"."id"
FROM 
  "people_db_attorney" 
INNER JOIN "people_db_role" ON 
  ("people_db_attorney"."id" = "people_db_role"."attorney_id" ) 
WHERE 
  ("people_db_role"."docket_id" = 7926834 AND "people_db_role"."party_id" IN (2374386, 2374387))
```

And the associated `EXPLAIN ANALYZE`:

```txt
HashAggregate  (cost=85991.88..86133.33 rows=14145 width=199) (actual time=101.115..105.202 rows=7398 loops=1)
   Group Key: people_db_role.party_id, people_db_attorney.id, people_db_attorney.date_created, people_db_attorney.date_modified, people_db_attorney.name, people_db_attorney.contact_raw, people_db_attorney.phone, people_db_attorney.fax, people_db_attorney.email
   ->  Gather  (cost=1502.49..85673.61 rows=14145 width=199) (actual time=18.760..82.259 rows=17069 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Nested Loop  (cost=502.49..83259.11 rows=5894 width=199) (actual time=14.780..84.386 rows=5690 loops=3)
               ->  Parallel Bitmap Heap Scan on people_db_role  (cost=502.06..39434.80 rows=5894 width=8) (actual time=14.491..61.882 rows=5690 loops=3)
                     Recheck Cond: (docket_id = 7926834)
                     Heap Blocks: exact=147
                     ->  Bitmap Index Scan on people_db_role_1427d4ab  (cost=0.00..498.52 rows=14145 width=0) (actual time=17.678..17.678 rows=17069 loops=1)
                           Index Cond: (docket_id = 7926834)
               ->  Index Scan using people_db_attorney_pkey on people_db_attorney  (cost=0.43..7.44 rows=1 width=195) (actual time=0.003..0.003 rows=1 loops=17069)
                     Index Cond: (id = people_db_role.attorney_id)
 Planning time: 0.625 ms
 Execution time: 106.324 ms
```

And visually:

![screenshot from 2019-01-03 22-49-17](https://user-images.githubusercontent.com/236970/50676556-d2151f00-0fa9-11e9-8436-831d8fafb433.png)

(Also [available here][good])

That should be looking a little better! But how do we get from bad to good with the ORM writing this query for us?

-----------------

To get at the heart of that question, we need to consider the Python that's doing this. It's nasty, but you can ignore most of it:

    ds = Docket.objects.filter(pk=d_pk).select_related(
        'originating_court_information',
    ).prefetch_related(
        'panel',
        'parties__party_types__criminal_complaints',
        'parties__party_types__criminal_counts',
        Prefetch(
            'docket_entries__recap_documents',
            queryset=RECAPDocument.objects.all().defer('plain_text')
        ),
        Prefetch(
            # Only attorneys in the docket.
            'parties__attorneys',
            # This is exactly where the problem happens! 
            queryset=Attorney.objects.filter(
                roles__docket_id=d_pk,
            ).distinct().prefetch_related(
                Prefetch(
                    'roles',
                    queryset=Role.objects.filter(docket_id=d_pk)
                )
            ),
        ),
    )

The issue is that in the queryset that pulls in the `Attorney` objects, we filter to the same table as the one that is used by the `Prefetch` object (`roles__docket_id=d_pk`). *That* creates the double join. 

Solutions were many:

 - Create a postgresql view of the table that we need, and then line it up with an unmanaged Django model. Then, simply query the model. 

    This might have worked, but I got fed up when I realized the Django model would be an entire duplication of other models. I also wasn't thrilled about creating the view and having to maintain it, migrate it, etc. (Though [this solution][solution] looked helpful.)

 - Use a custom Django Manager. Might have worked. Didn't feel right. Decided to move on.

 - Use Django `Subquery` objects. These were released for Django 1.9, but [backported][bp] to 1.8 (which we're on). 

    I think in theory these could have worked, but I never quite grokked how they'd do what we wanted. The general idea is that you'd do a subquery ahead of time and then it'd get `annotate`d onto the usual query. Seemed nice, but complicated. Also requires an additional dependency, though it's recommended by the guy that wrote the Django code for Subqueries, so it's probably...fine. Anyway, I didn't figure this out.

 - Redo the entire DB schema. Nope.

 - Use a raw SQL query. Seemed possible. Maybe it could have worked, but it comes with a list of cons a mile long.

 - Manually run the `Attorney` query, and prefetch and merge it ourselves in Python. This was getting desperate, but when started looking into this I discovered [a ticket][at] in Django's bug tracker with the following title:

    >  Make prefetch_related_objects a public API 

    Interesting, and sure enough in Django 1.9 or 1.10, there's a new method called `prefetch_related_objects` that allows you to do the merging yourself. It's even [documented][docs], and sure enough it existed in Django 1.8, though it was undocumented and a bit hidden. 

------

So. This has become longer than most blog posts, but I think we're in business using the last approach as in 3fcdf9de680261a1e69de53cb7306f4bfd7f4c92. The final approach still has some not great properties:

1. The double join is still there when you look at the queries, but because we're able to access the parties before we make our Prefetch query, we can put those IDs directly into the regular join as well. According to testing, this makes it perform well, unlike the other double-join query.

2. We now make one more query per docket since we have to fetch the parties twice. The first time is within the original `prefetch_related` method, and then the second is in the `prefetch_related_objects` call. 

Still, all that aside, I'm going to give it a spin. We'll see if it works and how well. Let's hope this is a permanent fix because I'm low on ideas (and out of time) otherwise.

[bad]: https://explain.depesz.com/s/lVPv
[good]: https://explain.depesz.com/s/KbpK
[solution]: https://stackoverflow.com/a/29532034/64911
[bp]: https://github.com/schinckel/django-subquery/
[at]: https://code.djangoproject.com/ticket/25279
[docs]: https://docs.djangoproject.com/en/2.1/ref/models/querysets/#prefetch-related-objects

**Comment by mlissner** [comment]
OK, I believe this is fixed and our uploads are flowing again. Fun bonus, I made them go from the replica, so they won't impact our performance.
